lib/mbedtls: Update to mbedtls v2.28.1.

Carglglz opened the pull request:
Update following recommendation #11355, which seems like a `esp-idf` <= `v4.2` deprecation is needed too (see [mbed-tls-support-in-esp-idf ](https://github.com/espressif/mbedtls/wiki#mbed-tls-support-in-esp-idf)) . Not sure how to update CI test suite to drop the `esp-idfv4.0.2` build? 🤔

Carglglz committed:
lib/mbedtls: Update to mbedtls v2.28.1.

Signed-off-by: Carlos Gil <carlosgilglez@gmail.com>

dpgeorge commented:
Thanks for doing this!

The esp32 IDF 4.0.2 CI is passing so that's good.  The changes here should not affect the esp32 port at all because it uses it's own version of mbedtls (not the submodule here), and also its own build scripts.

Carglglz commented:
Yes but `esp-idfv4.2` uses `mbedtls-v.2.16.2` and it seems it has no LTS (no updates since Dec 2021 [mbedtls @ dc406b8](https://github.com/espressif/mbedtls/tree/dc406b82b20481bde74a60fa8a776285d3fb9cf3))? so it may be a good idea to drop `esp-idf` <= `v4.2` i.e. only support `v4.3.x` and `v4.4.x` as I said...

dpgeorge commented:
Oh, I see what you mean.  Not that the code is now broken, just that it's using an obsolete library.

Yes, one thing I wanted to do before the next release is drop support for IDF less than 4.4.  That will allow cleaning up some old code in the esp32 port.  But that doesn't have to be done right now (it requires enabling `MICROPY_GC_SPLIT_HEAP` on esp32).

dpgeorge commented:
I think this might require updating the code in `lib/mbedtls_errors`.  Running `do-mp.sh` in that directory fails after this update.

robert-hh commented:
Oops. You're right, thanks. But that move might still be a temporary change. because we did not settle on the pin state while PWM started and after it has been stopped. And there is still the question, if there should be a start/stop set of methods. At the moment, deinit()/init() act as stop/start. But that's less clear.

Carglglz commented:
> mbedtls-2.28 — long-time support branch, only getting bug fixes. We plan to support this branch until December 2024. APIs are fully compatible with 2.16.

Does not this mean there is no change in error codes?, this is what I get
```diff
--- generate_errors_orig.pl     2020-06-20 08:40:38.819060379 -0700
+++ generate_errors.pl  2020-06-20 08:47:26.511163591 -0700
@@ -162,16 +162,12 @@

     if ($error_name eq "MBEDTLS_ERR_SSL_FATAL_ALERT_MESSAGE")
     {
-        ${$code_check} .= "${white_space}if( use_ret == -($error_name) )\n".
-                          "${white_space}\{\n".
-                          "${white_space}    mbedtls_snprintf( buf, buflen, \"$module_name - $description\" );\n".
-                          "${white_space}    return;\n".
-                          "${white_space}}\n"
+        # no-op, this case is hard-coded in error.fmt
     }
     else
     {
-        ${$code_check} .= "${white_space}if( use_ret == -($error_name) )\n".
-                          "${white_space}    mbedtls_snprintf( buf, buflen, \"$module_name - $description\" );\n"
+         my $error_text = $error_name =~ s/^MBEDTLS_ERR_//r;
+         ${$code_check} .= "${white_space}{ -($error_name), \"$error_text\" },\n"
     }
 };
```
I checked the diff between both `mbedtls/library/error.c` versions and error codes are the same, only restructured code as far as I can see. i.e. no api changes, however it looks like `mp_generate_errors.pl` parses `error.c` and expects the old structure. Not sure how to fix that (perl?). So I would expect that even when `mp_generate_errors.pl` is fixed the output in `mp_mbedtls_errors.c` should remain the same? 🤔

dpgeorge committed:
lib/mbedtls_errors: Update patch and error list for new mbedtls.

Running `./do-mp.sh` now generates this mp_mbedtls_errors.c with mbedTLS
v2.28.1.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge commented:
The error messages have changed slightly.  I push a fix to master (which should have been done when mbedtls was updated to v2.16), rebased this branch and then added a commit which fixes the Perl patch.  Running `./do-mp.sh` now passes, as does `./do-test.sh` and `./tester`.

dpgeorge commented:
With the updated error messages, IDF 4.0.2 and 4.4 do not build...

Is it worth continuing down this path (will require a separate `mp_mbedtls_errors.c` file for esp32), or going straight for mbedtls v3.x?

Carglglz commented:
> The error messages have changed slightly. I push a fix to master (which should have been done when mbedtls was updated to v2.16), rebased this branch and then added a commit which fixes the Perl patch. Running `./do-mp.sh` now passes, as does `./do-test.sh` and `./tester`.

Nice, I see now there are a few new errors, that's why builds failed.

> With the updated error messages, IDF 4.0.2 and 4.4 do not build...

I've just checked out `esp-idf` to see where exactly the `mbedtls-2.28` update was introduced.
`v4.3.3` and `v4.4.1` are the minimum version that support `mbedtls-2.28`. 
I see now that I used the `release/v4.4` which is `v4.4.4` and is not the same as tag `v4.4`.

So updating build_idf44 to use `v4.4.1` should fix that.
For the `idf-v4.0.2` build I only see 
> so it may be a good idea to drop esp-idf <= v4.2 i.e. only support v4.3.x and v4.4.x as I said

as an option.

> Is it worth continuing down this path (will require a separate mp_mbedtls_errors.c file for esp32), or going straight for mbedtls v3.x?

Updating to `mbedtls v3.x` will need to drop `esp-idf` <= `v5.0` , that or add a lot of `#macros`  to support both `mbedtls-2.28` and `mbedtls-v3.2.x` version (not sure if worth it? 😅 )

dpgeorge commented:
> So updating build_idf44 to use `v4.4.1` should fix that.

Ah, we should eventually update that to v4.4.4, which is the latest.  That should build fine (at least I know v4.4.3 does, it's what the daily firmware is built with).

> For the `idf-v4.0.2` build I only see [dropping it] as an option.

Yes I agree, we can drop support for anything < v4.4.1.  But that's a bit more involved (as mentioned above) and should be decoupled from the work in this PR.

> what about adding macros to `mp_mbedtls_error.c` to check `mbedtls version`?
> I think it would fix the problem right now.

That's quite tricky because `mp_mbedtls_errors.c` is auto-generated (by `do-mp.sh`), so the auto-generator (the Perl script) needs to be modified to insert macros to check the version.

Instead I think it's easiest to just provide a separate `esp32_mbedtls_errors.c` file, which is exactly what `do-esp32.sh` does if you run it.  Then update `ports/esp32/main/CMakeLists.txt` to point to this file instead of `mp_mbedtls_errors.c`.

Summary:
- create an `esp32_mbedtls_errors.c` file that works for both IDF v4.0.2 and v4.4 (no need to update the IDF version in the CI)
- apply the rest of this PR so all other ports use mbedtls v2.28.1 with an updated `mp_mbedtls_errors.c` file
- in the future remove support for IDF < v4.4.1 and potentially remove `esp32_mbedtls_errors.c`

Carglglz committed:
lib/mbedtls_errors: Add esp32-specific mbedtls error file.

This allows updating mp_mbedtls_errors.c for the other mbedtls based ports
based on mbedTLS v2.28.1.  This esp32-specific file will not be required
after updating IDF support to >= v4.4.1.

Signed-off-by: Carlos Gil <carlosgilglez@gmail.com>

dpgeorge commented:
Thanks for updating.  I reordered the commits and merged.
